### PR TITLE
start main.c

### PR DESCRIPTION
**Pull request by Cofflag**

### CODE REVIEW

File: src/main.c [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+extern void sub_08001C30(void);
+
+extern int gUnk_03002968;
+
+extern int* gUnk_03002970[5];
+extern int gUnk_08BDA40C;
+extern int gUnk_098C8D9C;
+extern int gUnk_09718FFC;
+extern int gUnk_094D446C;
+extern int gUnk_09E2261C;
```
Comment (jiangzhengwenjz) [inline src/main.c:17]:
```suggestion
extern const int gUnk_08BDA40C[];
extern const int gUnk_098C8D9C[];
extern const int gUnk_09718FFC[];
extern const int gUnk_094D446C[];
extern const int gUnk_09E2261C[];
```
Data in ROM is always const. And clearly they are arrays.

File: src/main.c [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    if (arr0[10])
+        arr1->unk14 = &arr0[arr0[10]>>2];
+    else
+        arr1->unk14 = NULL;
+    if (arr0[11]) {
+        arr1->unk18 = &arr0[arr0[11]>>2];
+        arr1->unka = (arr1->unk18[0]+1) * 4;
+    }
+    else {
+        arr1->unk18 = 0;
```
Comment (jiangzhengwenjz) [inline src/main.c:55]:
```suggestion
        arr1->unk18 = NULL;
```

File: src/main.c [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+extern void sub_08001B20(void);
+extern void sub_08001B14(void);
+extern void sub_08001A1C(void);
+extern void sub_08001AEC(int);
+extern void sub_08001C30(void);
+
+extern int gUnk_03002968;
+
+extern int* gUnk_03002970[5];
```
Comment (jiangzhengwenjz) [inline src/main.c:12]:
```suggestion
extern const int *gUnk_03002970[5];
```
See below.

File: src/main.c [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+}
+//TODO:type
+
+void sub_08001D0C(void)
+{
+    gUnk_03002970[0] = &gUnk_08BDA40C;
+    gUnk_03002970[1] = &gUnk_098C8D9C;
+    gUnk_03002970[2] = &gUnk_09718FFC;
+    gUnk_03002970[3] = &gUnk_094D446C;
+    gUnk_03002970[4] = &gUnk_09E2261C;
```
Comment (jiangzhengwenjz) [inline src/main.c:67]: `&` can be removed since they are changed to array names now.

File: src/main.c [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    int *var = &a[b * 2 + 2];
+    return &a[*var * 4];
+}
+
+int sub_08001D88(u32 *a,u16 b)
+{
+    //u32 *var1 = &a[b*2];
+    //return var1[3]*16;
+    //registers will swap.
+    return a[b*2+3]*16;
```
Comment (jiangzhengwenjz) [inline src/main.c:95]: use space please

**Review by jiangzhengwenjz — CHANGES_REQUESTED** [review]

File: src/main.c [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    gUnk_03002970[0] = &gUnk_08BDA40C;
+    gUnk_03002970[1] = &gUnk_098C8D9C;
+    gUnk_03002970[2] = &gUnk_09718FFC;
+    gUnk_03002970[3] = &gUnk_094D446C;
+    gUnk_03002970[4] = &gUnk_09E2261C;
+}
+
+int *sub_08001D3C(u16 a, u16 b)
+{
+    int *var = &gUnk_03002970[a][b * 2 + 2];
```
Comment (jiangzhengwenjz) [inline src/main.c:72]:
```suggestion
    const int *var = &gUnk_03002970[a][b * 2 + 2];
```

File: src/main.c [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    return &gUnk_03002970[a][*var * 4];
+}
+//Compiler will write lsl/lsr automatically if arguments are u16s. 
+
+int sub_08001D5C(u16 a,u16 b)
+{
+    int* var1 = &gUnk_03002970[a][b*2];
```
Comment (jiangzhengwenjz) [inline src/main.c:79]:
```suggestion
    const int *var1 = &gUnk_03002970[a][b*2];
```

File: src/main.c [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+int *sub_08001D3C(u16 a, u16 b)
+{
+    int *var = &gUnk_03002970[a][b * 2 + 2];
+    return &gUnk_03002970[a][*var * 4];
+}
+//Compiler will write lsl/lsr automatically if arguments are u16s. 
+
+int sub_08001D5C(u16 a,u16 b)
```
Comment (jiangzhengwenjz) [inline src/main.c:77]:
```suggestion
int sub_08001D5C(u16 a, u16 b)
```

File: src/main.c [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    //registers will swap.
+    return a[b*2+3]*16;
+}
+
+u16 sub_08001D94(u16 *a)
+{
+    return a[0];
+}
+
+void sub_08001D98(u32 *a,struct unk_2 *b)
```
Comment (jiangzhengwenjz) [inline src/main.c:103]:
```suggestion
void sub_08001D98(u32 *a, struct unk_2 *b)
```

File: src/main.c [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    b->unk8 = &a[var];
+    
+    var = a[7]/4;
+    b->unkc = &a[var];
+
+    var = a[9]/4;
+    b->unk10 = &a[var];   
+}
+
+void sub_08001DC4(struct unk_3 *a,struct unk_4 *b)
```
Comment (jiangzhengwenjz) [inline src/main.c:121]:
```suggestion
void sub_08001DC4(struct unk_3 *a, struct unk_4 *b)
```

File: src/main.c [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+    b->unk0 = a->unk10;
+    b->unk2 = a->unk12;
+    b->unk4 = a->unk14;
+    b->unk6 = a->unk16;
+    b->unk8 = a->unk18;
+    b->unka = a->unk1a;
+    b->unkc = a->unk0 + a->unk1c/4;
+}
+
+void sub_08001DE8(struct unk_5 *a,struct unk_6 *b)
```
Comment (jiangzhengwenjz) [inline src/main.c:132]:
```suggestion
void sub_08001DE8(struct unk_5 *a, struct unk_6 *b)
```

File: src/main.c [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+    while (1)
+    {
+        sub_08001A1C();
+        sub_08001AEC(1);
+        sub_08001C30();
+    }
+
+}
+
+void sub_08001CBC(u32* arr0,struct unk_1* arr1)
```
Comment (jiangzhengwenjz) [inline src/main.c:36]:
```suggestion
void sub_08001CBC(u32* arr0, struct unk_1* arr1)
```

File: src/main.c [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+    return 16*gUnk_03002970[a][b*2+3];
+}
+
+int *sub_08001D78(int *a, u16 b)
+{
+    int *var = &a[b * 2 + 2];
+    return &a[*var * 4];
+}
+
+int sub_08001D88(u32 *a,u16 b)
```
Comment (jiangzhengwenjz) [inline src/main.c:90]:
```suggestion
int sub_08001D88(u32 *a, u16 b)
```

**Review by jiangzhengwenjz — CHANGES_REQUESTED** [review]

File: src/main.c [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+    b->unk0 = a[8];
+    b->unk4 = a + 9;
+}
+
+s16 sub_08001E70(u16 a) 
+{
+    if(a > 359)
+    {
+        a = DivRem(a,0xb4 * 2);
+    }
```
Comment (jiangzhengwenjz) [inline src/main.c:154]:
```suggestion
    if(a >= 360)
        a = DivRem(a, 360);
```

File: src/main.c [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+}
+
+s16 sub_08001ED4(s16 a, s16 b) 
+{
+    return a * b >> 8;
+}
+
+s16 sub_08001EE4(s16 a, s16 b) 
+{
+    return Div(a << 8, b);
```
Comment (jiangzhengwenjz) [inline src/main.c:174]: Is `a * 0x100` matching? `a << 8` is undefined behavior if `a` is negative.

File: src/main.c [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+}
+
+s16 sub_08001EFC(s16 a) 
+{
+    s32 var1 = 65536;
```
Comment (jiangzhengwenjz) [inline src/main.c:179]:
```suggestion
    s32 var1 = 0x10000;
```

**Review by jiangzhengwenjz — CHANGES_REQUESTED** [review]

File: src/main.c (hunk 15 above)
Comment (Cofflag) [reply src/main.c:174]: Yes

File: src/main.c [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+    gUnk_03002970[4] = gUnk_09E2261C;
+}
+
+const int *sub_08001D3C(u16 a, u16 b)
+{
+    const int *var = &gUnk_03002970[a][b * 2 + 2];
+    return &gUnk_03002970[a][*var * 4];
+}
+
+const int sub_08001D5C(u16 a, u16 b)
```
Comment (jiangzhengwenjz) [inline src/main.c:57]:
```suggestion
int sub_08001D5C(u16 a, u16 b)
```

**Review by jiangzhengwenjz — CHANGES_REQUESTED** [review]

File: src/main.c [hunk 18]
```diff
@@ -4,6 +4,10 @@
 #include "data.h"
 #include "functions.h"
 
+int sub_08001A18(int *a)
+{
+    return *a;
+}
```
Comment (jiangzhengwenjz) [inline src/main.c:10]: need extra newline between functions

File: src/main.c [hunk 19]
```diff
@@ -1,19 +1,168 @@
...
+
+void sub_08001B20(void)
+{
+    REG_WAITCNT = WAITCNT_SRAM_4 | WAITCNT_WS0_N_3 | WAITCNT_WS0_S_1 | WAITCNT_PREFETCH_ENABLE;
+
+    DmaFill32(3, 0x55555555, (void*) EWRAM_START, 0x40000);
+    DmaFill32(3, 0x55555555, (void*) IWRAM_START, 0x7b00);
+    DmaFill32(3, 0, (void*) VRAM, 0x18000);
+    DmaFill32(3, 0xa0, (void*) OAM, 0x400);
+    DmaFill16(3, 0, (void*) PLTT, 0x400);
```
Comment (jiangzhengwenjz) [inline src/main.c:95]: I believe you don't need the void * cast

File: src/main.c [hunk 20]
```diff
@@ -1,19 +1,168 @@
...
+    sub_08012F0C(0,0x55555555);
+    sub_08012F0C(1,0x55555555);
+    sub_08012F0C(2,0x55555555);
+    sub_08012F0C(3,0x55555555);
+}
+
+void sub_08001AEC(int unk)
+{
+    gUnk_03002960 = (s16)unk;
+}
```
Comment (jiangzhengwenjz) [inline src/main.c:70]:
```c
void sub_08001AEC(s16 unk)
{
    gUnk_03002960 = unk;
}
```

File: src/main.c [hunk 21]
```diff
@@ -148,19 +297,19 @@ void sub_08001E64(u16 *a, struct unk_11 *b)
 
 s16 sub_08001E70(u16 a) 
 {
-    if(a > 359)
-    {
-        a = DivRem(a,0xb4 * 2);
-    }
+    if(a >= 360)
```
Comment (jiangzhengwenjz) [inline src/main.c:304]: space between `if` and `(`

File: src/main.c [hunk 22]
```diff
@@ -148,19 +297,19 @@ void sub_08001E64(u16 *a, struct unk_11 *b)
...
     return gUnk_08B6D24C[a];
 }
 
 s16 sub_08001EA0(u16 a) 
 {
-    if(a > 359)
-    {
-        a = DivRem(a,0xb4 * 2);
-    }
+    if(a >= 360)
```
Comment (jiangzhengwenjz) [inline src/main.c:313]: space between `if` and `(`

File: src/main.c [hunk 23]
```diff
@@ -148,19 +297,19 @@ void sub_08001E64(u16 *a, struct unk_11 *b)
...
 s16 sub_08001E70(u16 a) 
 {
-    if(a > 359)
-    {
-        a = DivRem(a,0xb4 * 2);
-    }
+    if(a >= 360)
+
+        a = DivRem(a, 360);
+
```
Comment (jiangzhengwenjz) [inline src/main.c:307]: why do you need extra newlines here?

File: src/main.c [hunk 24]
```diff
@@ -148,19 +297,19 @@ void sub_08001E64(u16 *a, struct unk_11 *b)
...
 s16 sub_08001EA0(u16 a) 
 {
-    if(a > 359)
-    {
-        a = DivRem(a,0xb4 * 2);
-    }
+    if(a >= 360)
+
+        a = DivRem(a, 360);
+
```
Comment (jiangzhengwenjz) [inline src/main.c:316]: why do you need extra newlines here?

**Review by jiangzhengwenjz — CHANGES_REQUESTED** [review]

File: src/main.c [hunk 25]
```diff
@@ -0,0 +1,334 @@
...
+        sub_08019688();
+        sub_08001C00();
+        sub_08001C64();
+        VBlankIntrWait();
+    }
+}
+
+void sub_08001C64(void)
+{
+    if ( gUnk_03002964 == 1 && (gUnk_0300594C & 0xF) == 15 )
```
Comment (jiangzhengwenjz) [inline src/main.c:147]:
```suggestion
    if (gUnk_03002964 == 1 && (gUnk_0300594C & 0xF) == 0xF)
```

File: src/main.c [hunk 26]
```diff
@@ -0,0 +1,334 @@
...
+#include "data.h"
+#include "functions.h"
+
+int sub_08001A18(int *a)
+{
+    return *a;
+}
+void sub_08001A1C(void)
+{
+    int i;
```
Comment (jiangzhengwenjz) [inline src/main.c:13]: You can remove this variable.

File: src/main.c [hunk 27]
```diff
@@ -0,0 +1,334 @@
...
+    sub_080150C4();
+    sub_08006AA4();
+    sub_08004A8C();
+    sub_080092B0();
+    sub_0800A5BC();
+    sub_0801978C();
+    sub_080AB2D0();
+    
+    i = sub_08012F60(0x180);
+    if (i != 1)
```
Comment (jiangzhengwenjz) [inline src/main.c:27]:
```suggestion
    if (sub_08012F60(0x180) != 1)
```

File: src/main.c [hunk 28]
```diff
@@ -0,0 +1,334 @@
...
+    sub_0800F094();
+    sub_080136CC();
+    sub_08003620();
+    sub_0800AF74();
+    sub_0801B960();
+    sub_08008B3C();
+    sub_080637C4();
+    
+    i = sub_08012F60(0x180);
+    if (i != 1)
```
Comment (jiangzhengwenjz) [inline src/main.c:41]:
```suggestion
    if (sub_08012F60(0x180) != 1)
```

File: src/main.c [hunk 29]
```diff
@@ -0,0 +1,334 @@
...
+        sub_08017B1C();
+    }
+
+    sub_0806EF50();
+    sub_080B3128();
+    sub_080B4970();
+    sub_08062EB0();
+    
+    i = sub_08012F60(0x180);
+    if (i != 1)
```
Comment (jiangzhengwenjz) [inline src/main.c:52]:
```suggestion
    if (sub_08012F60(0x180) != 1)
```

File: src/main.c [hunk 30]
```diff
@@ -0,0 +1,334 @@
...
+
+void sub_08001E64(u16 *a, struct unk_11 *b) 
+{
+    b->unk0 = a[8];
+    b->unk4 = a + 9;
+}
+
+s16 sub_08001E70(u16 a) 
+{
+    if(a >= 360)
```
Comment (jiangzhengwenjz) [inline src/main.c:304]: space between `if` and `(`

**Review by jiangzhengwenjz — CHANGES_REQUESTED** [review]

File: src/main.c [hunk 31]
```diff
@@ -0,0 +1,326 @@
...
+    }
+
+    sub_08012D14(0,&gUnk_02006000);
+    sub_08012F60(4);
+    sub_08012D30();
+    sub_08012D64(0,0,0);
+    sub_08012F0C(0,0x55555555);
+    sub_08012F0C(1,0x55555555);
+    sub_08012F0C(2,0x55555555);
+    sub_08012F0C(3,0x55555555);
```
Comment (jiangzhengwenjz) [inline src/main.c:60]: space after `,`

File: src/main.c [hunk 32]
```diff
@@ -0,0 +1,326 @@
...
+        sub_08019688();
+        sub_08001C00();
+        sub_08001C64();
+        VBlankIntrWait();
+    }
+}
+
+void sub_08001C64(void)
+{
+    if (gUnk_03002964 == 1 && (gUnk_0300594C & 0xF) == 0xF )
```
Comment (jiangzhengwenjz) [inline src/main.c:143]:
```suggestion
    if (gUnk_03002964 == 1 && (gUnk_0300594C & 0xF) == 0xF)
```

**Review by jiangzhengwenjz — CHANGES_REQUESTED** [review]

**Review by jiangzhengwenjz — APPROVED** [review]

**Pull request merged**